### PR TITLE
[SPARK-23846][SQL] The samplingRatio option for CSV datasource

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

I propose to support the `samplingRatio` option for schema inferring of CSV datasource similar to the same option of JSON datasource: 
https://github.com/apache/spark/blob/b14993e1fcb68e1c946a671c6048605ab4afdf58/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala#L49-L50

## How was this patch tested?

Added 2 tests for json and 2 tests for csv datasources. The tests checks that only subset of input dataset is used for schema inferring.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 1]
```diff
@@ -2127,4 +2127,39 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
       assert(df.schema === expectedSchema)
     }
   }
+
+  test("schema inferring touches less data if samplingRation < 1.0") {
```
Comment (sujithjay) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: Hi @MaxGekk , since these test-cases in `JsonSuite` are not related to the changes made in this patch, can it be raised as a separate PR?

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: @sujithjay I created separate PR for the tests: https://github.com/apache/spark/pull/20963

**Comment by HyukjinKwon** [comment]
_if I remember this correctly_, there was a discussion about it a lone ago and, @rxin was not sure how much it improves the perf and if it's worth, which I ended up with agreeing with. @rxin, did I recall this correctly?

There's a workaround for it BTW:

```scala
val ds = Seq("a", "b", "c", "d").toDS
val sampledSchema = spark.read.option("inferSchema", true).csv(ds.sample(false, 0.7)).schema
spark.read.schema(sampledSchema).csv(ds)
```

**Comment by MaxGekk** [comment]
@rxin I made an experiment on json files but numbers for csv are almost the same. For example, inferring schema for 50GB json:
```
scala> spark.read.option("samplingRatio", 0.000000001).json("test.json")
```
took 1.7 minute
```
scala> spark.read.option("samplingRatio", 1.0).json("test.json")
```
took 21.9 minutes.

I have looked in a profiler where Spark spends time during schema inferring for 50GB json. At least on my laptop - 75% in json parsing and 18% on disk IO. Of course, the numbers will be different in a cluster if the files would be read from s3 via network. In any way, the samplingRatio option gives us opportunity to find a balance of CPUs load and network/disk IO. 

@HyukjinKwon The question is not about workaround, it is about usability:

1. For interactive queries, an user doesn't have to write the boilerplate code if there is the option.

2. If the code is used inside of a library, developers don't have to check special cases like if it is json use the samplingRatio option otherwise do sampling manually.

Additionally the behavior behind of the option could be improved in the future. For example, it will require less file reads during sampling. It would be easer to do that with the option.

**Comment by HyukjinKwon** [comment]
For usability, the workaround I suggested above has more flexibility. For example, we can make different operation (e.g, filter) on schema inference path. They are only few lines.

Schema inference is discouraged in production line. I believe, for example, just taking 100 records and use the schema makes more sense.

I am not against this option but I am saying I don't feel strong on this for above reasons.

**Comment by rxin** [comment]
I'm good with having this option given the data @MaxGekk posted. (I haven't reviewed the code - somebody else should do that before merging).

`val sampledSchema = spark.read.option("inferSchema", true).csv(ds.sample(false, 0.7)).schema` is a bit clunky compared with an option that applies to all the sources.

**Comment by HyukjinKwon** [comment]
sure, will review this and go ahead.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 2]
```diff
@@ -1279,4 +1279,45 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
       Row("0,2013-111-11 12:13:14") :: Row(null) :: Nil
     )
   }
+
+  test("SPARK-23846: schema inferring touches less data if samplingRation < 1.0") {
+    val predefinedSample = Set[Int](2, 8, 15, 27, 30, 34, 35, 37, 44, 46,
+      57, 62, 68, 72)
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: `val predefinedSample = Set[Int](2, 8, 15, 27, 30, 34, 35, 37, 44, 46)` is enough.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 3]
```diff
@@ -1279,4 +1279,45 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
+        57, 62, 68, 72)
+      if (predefinedSample.contains(i)) {
+        i.toString + "\n"
+      } else {
+        (i.toDouble + 0.1) + "\n"
+      }
+    }.toDS()
+    val ds = spark.read
+      .option("inferSchema", true)
+      .option("samplingRatio", 0.1)
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Add some negative case.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 4]
```diff
@@ -528,6 +529,7 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
    * <li>`header` (default `false`): uses the first line as names of columns.</li>
    * <li>`inferSchema` (default `false`): infers the input schema automatically from data. It
    * requires one extra pass over the data.</li>
+   * <li>`samplingRatio` (default 1.0): the sample ratio of rows used for schema inferring.</li>
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Also need to update the PySpark API

**Comment by gatorsmile** [comment]
@MaxGekk Thanks for working on this!

**Comment by HyukjinKwon** [comment]
Let's address @gatorsmile's and mine at https://github.com/apache/spark/pull/20963 too as well. Seems fine otherwise.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Sampling of 100 elements return 13 elements, if it is called with `withReplacement` = `false` and `seed` = `1` . The test fails if I select only 10 elements. I took the same values for  `withReplacement` and `seed` as in JSON datasource.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 4 above)
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: I just did absolutely the same as `samplingRatio` for JSON datasource - `samplingRatio` is not supported as a parameter of `json()` but could be set as the option like  `.option('samplingRatio', '0.1')`.  Should I update PySpark API regarding `samplingRatio` for `json()` in separate PR?

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Yes, please. Seems this option can be useful given your investigation and @rxin's opinion.

File: python/pyspark/sql/tests.py [hunk 5]
```diff
@@ -3009,6 +3009,15 @@ def test_sort_with_nulls_order(self):
             df.select(df.name).orderBy(functions.desc_nulls_last('name')).collect(),
             [Row(name=u'Tom'), Row(name=u'Alice'), Row(name=None)])
 
+    def test_csv_sampling_ratio(self):
+        rdd = self.spark.sparkContext.range(0, 100).\
+            map(lambda x: '0.1' if x == 1 else str(x)).\
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: `.\nmap` -> `\n.map`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 6]
```diff
@@ -1279,4 +1278,68 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
+        .option("inferSchema", true)
+        .option("samplingRatio", 0.1)
+        .csv(path.getCanonicalPath)
+      assert(ds.schema == new StructType().add("_c0", IntegerType))
+    })
+  }
+
+  test("SPARK-23846: usage of samplingRatio while parsing a dataset of strings") {
+    val dstr = spark.sparkContext.parallelize(0 until 100, 1).map { i =>
+      val predefinedSample = Set[Int](2, 8, 15, 27, 30, 34, 35, 37, 44, 46,
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Can we take this out of `map`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 7]
```diff
@@ -1279,4 +1278,68 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
+        57, 62, 68, 72)
+      if (predefinedSample.contains(i)) {
+        i.toString + "\n"
+      } else {
+        (i.toDouble + 0.1) + "\n"
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Hm .. does it need `\n`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 8]
```diff
@@ -1279,4 +1278,68 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
+
+  test("SPARK-23846: schema inferring touches less data if samplingRatio < 1.0") {
+    // Set default values for the DataSource parameters to make sure
+    // that whole test file is mapped to only one partition. This will guarantee
+    // reliable sampling of the input file.
+    withSQLConf(
+      "spark.sql.files.maxPartitionBytes" -> (128 * 1024 * 1024).toString,
+      "spark.sql.files.openCostInBytes" -> (4 * 1024 * 1024).toString
+    )(withTempPath { path =>
+      val rdd = spark.sqlContext.range(0, 100).map {row =>
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Can we actually do

```
        val predefinedSample = Set[Long](2, 8, 15, 27, 30, 34, 35, 37, 44, 46,
          57, 62, 68, 72)
        val value = row.getLong(0)
        if (predefinedSample.contains(value)) {
          value.toString
        } else {
          (value.toDouble + 0.1).toString
        }
```

outside of `map`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 9]
```diff
@@ -1279,4 +1278,64 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
       Row("0,2013-111-11 12:13:14") :: Row(null) :: Nil
     )
   }
+
+  val sampledTestData = (row: Row) => {
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: I have to define it as `val` because `def` contains a reference to outer class `CSVSuite` and the lambda inside of `map(...)` becomes not serializable.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 4 above)
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Please, have a look at the PR: https://github.com/apache/spark/pull/21056

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala [hunk 10]
```diff
@@ -161,7 +161,8 @@ object TextInputCSVDataSource extends CSVDataSource {
       val firstRow = new CsvParser(parsedOptions.asParserSettings).parseLine(firstLine)
       val caseSensitive = sparkSession.sessionState.conf.caseSensitiveAnalysis
       val header = makeSafeHeader(firstRow, caseSensitive, parsedOptions)
-      val tokenRDD = csv.rdd.mapPartitions { iter =>
+      val sampled: Dataset[String] = CSVUtils.sample(csv, parsedOptions)
+      val tokenRDD = sampled.rdd.mapPartitions { iter =>
```
[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: No, it doesn't. The behavior is the same as in JSON datasource.

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: Doesn't JSON do the sampling when multiLine is enabled?

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]:
```
$ tree .
.
├── a.json
├── b.json
├── c.json
├── d.json
├── e.json
├── f.json
└── h.json

0 directories, 7 files
```

```
$ cat *
{"a": "a"}
{"a": 1}

{"a": 1}

{"a": 1}

{"a": 1}

{"a": 1}

{"a": 1}
```

```scala
scala> spark.read.option("samplingRatio", 0.1).json("tmp").schema
res0: org.apache.spark.sql.types.StructType = StructType(StructField(a,LongType,true))

scala> spark.read.option("samplingRatio", 1).json("tmp").schema
res1: org.apache.spark.sql.types.StructType = StructType(StructField(a,StringType,true))
```

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]:
```
$ tree .
.
├── a.json
├── b.json
├── c.json
├── d.json
├── e.json
├── f.json
└── h.json

0 directories, 7 files
```

```
$ cat *
{"a": "a"}
{"a": 1}

{"a": 1}

{"a": 1}

{"a": 1}

{"a": 1}

{"a": 1}
```

```scala
scala> spark.read.option("samplingRatio", 0.2).option("multiLine", true).json("tmp").schema
res0: org.apache.spark.sql.types.StructType = StructType(StructField(a,LongType,true))

scala> spark.read.option("samplingRatio", 1).option("multiLine", true).json("tmp").schema
res1: org.apache.spark.sql.types.StructType = StructType(StructField(a,StringType,true))
```

[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: Thank you @HyukjinKwon , I didn't know the `samplingRatio` option in the case of `multiLine` means sampling of files but not lines. I will add the implementation for CSV and update the PR https://github.com/apache/spark/pull/21056 by new tests. I guess we should explicitly describe the behavior in the comments of `samplingRatio`

File: python/pyspark/sql/readwriter.py [hunk 11]
```diff
@@ -882,6 +882,9 @@ def csv(self, path, mode=None, compression=None, sep=None, quote=None, escape=No
                                           the quote character. If None is set, the default value is
                                           escape character when escape and quote characters are
                                           different, ``\0`` otherwise..
+        :param samplingRatio: defines fraction of rows (when ``multiLine`` is ``false``) or fraction
+                              of files (when ``multiLine`` is ``true``) used for schema inferring.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/readwriter.py]: @MaxGekk, I think in CSV's case we shouldn't do sample based on the files .. JSON's case is a record per file whereas CSV allows multiple lines in a file ...

File: python/pyspark/sql/readwriter.py [hunk 12]
```diff
@@ -832,7 +832,7 @@ def text(self, path, compression=None, lineSep=None):
     def csv(self, path, mode=None, compression=None, sep=None, quote=None, escape=None,
             header=None, nullValue=None, escapeQuotes=None, quoteAll=None, dateFormat=None,
             timestampFormat=None, ignoreLeadingWhiteSpace=None, ignoreTrailingWhiteSpace=None,
-            charToEscapeQuoteEscaping=None):
+            charToEscapeQuoteEscaping=None, samplingRatio=None):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/readwriter.py]: This is in DataFrameWriter ....

File: python/pyspark/sql/tests.py [hunk 13]
```diff
@@ -3009,6 +3009,15 @@ def test_sort_with_nulls_order(self):
             df.select(df.name).orderBy(functions.desc_nulls_last('name')).collect(),
             [Row(name=u'Tom'), Row(name=u'Alice'), Row(name=None)])
 
+    def test_csv_sampling_ratio(self):
+        rdd = self.spark.sparkContext.range(0, 100)\
+            .map(lambda x: '0.1' if x == 1 else str(x))\
+            .repartition(1)
+        schema = self.spark.read.option('inferSchema', True)\
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: Let's test `.csv(samplingRatio=0.5, ...)`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 14]
```diff
@@ -19,23 +19,24 @@ package org.apache.spark.sql.execution.datasources.csv
...
 import org.apache.spark.SparkException
 import org.apache.spark.sql.{AnalysisException, DataFrame, QueryTest, Row, UDT}
 import org.apache.spark.sql.catalyst.util.DateTimeUtils
-import org.apache.spark.sql.functions.{col, regexp_replace}
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.test.{SharedSQLContext, SQLTestUtils}
 import org.apache.spark.sql.types._
 
-class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
+class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils  with TestCsvData {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: `  ` -> ` `.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 15]
```diff
@@ -1279,4 +1280,72 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
+      spark.read.option("inferSchema", true).option("samplingRatio", -1).csv(ds)
+    }.getMessage
+    assert(errorMsg0.contains("samplingRatio (-1.0) should be greater than 0"))
+
+    val errorMsg1 = intercept[IllegalArgumentException] {
+      spark.read.option("inferSchema", true).option("samplingRatio", 0).csv(ds)
+    }.getMessage
+    assert(errorMsg1.contains("samplingRatio (0.0) should be greater than 0"))
+
+    val sampled = spark.read.option("inferSchema", true).option("samplingRatio", 10).csv(ds)
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Can we leave this case out? Looks a bit odd we test with sampling ratio more than 1. Maybe we can just leave this case out here for now.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 16]
```diff
@@ -1279,4 +1280,72 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
+    assert(sampled.count() == ds.count())
+  }
+
+  test("SPARK-23846: sampling files for schema inferring in the multiLine mode") {
+    withTempDir { dir =>
+      Files.write(Paths.get(dir.getAbsolutePath, "0.csv"), "0.1".getBytes,
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: `getBytes` with explicit UTF-8?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 17]
```diff
@@ -1279,4 +1280,72 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
+      for (i <- 1 until 10) {
+        Files.write(Paths.get(dir.getAbsolutePath, s"$i.csv"), s"$i".getBytes,
+          StandardOpenOption.CREATE_NEW)
+      }
+      val files = (0 until 10).map { i =>
+        val hadoopConf = spark.sessionState.newHadoopConf()
+        val hdfsPath = new Path(Paths.get(dir.getAbsolutePath, s"$i.csv").toString)
+        hdfsPath.getFileSystem(hadoopConf).getFileStatus(hdfsPath)
+      }
+      // The test uses the internal method because public API cannot guarantee order of files
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: ditto - https://github.com/apache/spark/pull/21056#discussion_r183982653

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
LGTM

Merged to master.
